samanthas_telegram_bot/api_clients/base/base_api_client.py

```diff
@@ -0,0 +1,274 @@
+        data: DataDict | None = None,
+        json_data: DataDict | None = None,
+        params: DataDict | None = None,
+    ) -> tuple[int, DataDict]:
+        response = await cls._make_request_with_retries(
+            update=update,
+            context=context,
+            method=method,
+            url=url,
+            headers=headers,
+            data=data,
+            json_data=json_data,
+            params=params,
+        )
+
+        try:
+            status_code, json_data = cls._get_status_code_and_json(response)
+        except AttributeError as err:
+            raise BaseApiClientError("Could not load JSON") from err
+
+        try:
+            notification_params = notification_params_for_status_code[status_code]
+        except KeyError as err:
+            raise BaseApiClientError(
+                f"Unexpected {status_code=} after sending a {method} "
+                f"request to {url} with {data=}. JSON data received: {json_data}"
+            ) from err
+
+        await logs(
+            bot=context.bot,
+            level=notification_params.logging_level,
+            text=notification_params.message,
+            # API client creates one additional layer between caller and logger
+            stacklevel=CALLER_LOGGING_STACK_LEVEL + 1,
+            needs_to_notify_admin_group=notification_params.notify_admins,
+            parse_mode_for_admin_group_message=notification_params.parse_mode_for_bot_message,
+            update=update,
+        )
+
+        return status_code, json_data
+
+    @classmethod
+    async def _make_request_with_retries(
+        cls,
+        update: Update,
+        context: CUSTOM_CONTEXT_TYPES,
+        method: HttpMethod,
+        url: str,
+        headers: dict[str, str] | None = None,
+        data: DataDict | None = None,
+        json_data: DataDict | None = None,
+        params: DataDict | None = None,
+    ) -> Response:
+        attempts = 0
+        timeout = BASE_TIMEOUT_IN_SECS_BETWEEN_API_REQUEST_ATTEMPTS
+
+        while True:
+            try:
+                response = await cls._make_one_request(
+                    method=method,
+                    url=url,
+                    headers=headers,
+                    data=data,
+                    json_data=json_data,
+                    params=params,
+                )
+            except (NotImplementedError, httpx.TransportError) as err:
+                # TODO logic can be split here, maybe let user try again in case of httpx error.
+                raise BaseApiClientError(
+                    f"Failed to send {method.upper()} request to {url=}"
+                ) from err
+
+            if response.status_code not in (
+                httpx.codes.INTERNAL_SERVER_ERROR,
+                httpx.codes.BAD_GATEWAY,
+                httpx.codes.SERVICE_UNAVAILABLE,
+                httpx.codes.GATEWAY_TIMEOUT,
+            ):
+                return response
+
+            attempts += 1
```

Comment:
We should do some `max_attempts` checking here, otherwise it could be infinite retry loop

Reply:
Thanks for the catch, it got lost during reworking. Also fixed wrong positioning of attempts counter in SmallTalk client.